osu.Game.Rulesets.Osu/Difficulty/Evaluators/RhythmEvaluator.cs

```diff
@@ -73,33 +125,44 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current)
                         if (prevObj.BaseObject is Slider) // bpm change was from a slider, this is easier typically than circle -> circle
                             effectiveRatio *= 0.25;
 
-                        if (previousIslandSize == islandSize) // repeated island size (ex: triplet -> triplet)
-                            effectiveRatio *= 0.25;
-
-                        if (previousIslandSize % 2 == islandSize % 2) // repeated island polartiy (2 -> 4, 3 -> 5)
+                        if (previousIsland.Deltas.Count % 2 == island.Deltas.Count % 2) // repeated island polartiy (2 -> 4, 3 -> 5)
                             effectiveRatio *= 0.50;
 
-                        if (lastDelta > prevDelta + 10 && prevDelta > currDelta + 10) // previous increase happened a note ago, 1/1->1/2-1/4, dont want to buff this.
+                        if (lastDelta > prevDelta + deltaDifferenceEpsilon && prevDelta > currDelta + deltaDifferenceEpsilon) // previous increase happened a note ago, 1/1->1/2-1/4, dont want to buff this.
                             effectiveRatio *= 0.125;
 
-                        rhythmComplexitySum += Math.Sqrt(effectiveRatio * startRatio) * currHistoricalDecay * Math.Sqrt(4 + islandSize) / 2 * Math.Sqrt(4 + previousIslandSize) / 2;
+                        if (islandCounts.ContainsKey(island))
+                        {
+                            islandCounts[island]++;
+
+                            // repeated island (ex: triplet -> triplet)
+                            double power = Math.Max(0.75, logistic(island.AverageDelta(), 3, 0.15, 9));
+                            effectiveRatio *= Math.Pow(1.0 / islandCounts[island], power);
```

Comment:
we might want to scale this harsher since the common offenders of island repetition aren't really nerfed as much as I'd like